pkg/image/cdi/cdi.go

```diff
@@ -1,32 +1,269 @@
+	}
+	if apierrors.IsNotFound(err) {
+		logrus.Info("DataVolume not found, waiting for the initialization")
+		return err
+	}
+
+	// upload source type will update the progress on the upload handler
+	if vmImg.Spec.SourceType == harvesterv1.VirtualMachineImageSourceTypeDownload ||
+		vmImg.Spec.SourceType == harvesterv1.VirtualMachineImageSourceTypeExportVolume {
+		progress := string(targetDV.Status.Progress)
+		if progress != "N/A" && progress != "" {
+			// progress format looks like "88.82%", we just need the integer part
+			parsedInt := strings.Split(progress, ".")[0]
+			progressInt, err := strconv.Atoi(parsedInt)
+			if err != nil {
+				return fmt.Errorf("failed to convert progress to int: %v", err)
+			}
+			logrus.Infof("Update CDI DataVolume progress: %v", progressInt)
+			b.vmio.Importing(vmImg, "Image Importing", progressInt)
+		}
+		logrus.Infof("CDI DataVolume %s/%s status: %s, progress: %v", targetDVNs, targetDVName, targetDV.Status.Phase, progress)
+	}
+	if targetDV.Status.Phase != cdiv1.Succeeded {
+		return common.ErrRetryLater
+	}
+	b.vmio.Imported(vmImg, "", 100, vmImg.Status.Size, vmImg.Status.VirtualSize)
 	return nil
 }
 
-func (cb *Backend) UpdateVirtualSize(vmi *harvesterv1.VirtualMachineImage) (*harvesterv1.VirtualMachineImage, error) {
+func (b *Backend) UpdateVirtualSize(vmi *harvesterv1.VirtualMachineImage) (*harvesterv1.VirtualMachineImage, error) {
+	// CDI backend does not need this function, the VirtualSize is updated before the DataVolume is created
 	return vmi, nil
 }
 
-func (cb *Backend) Delete(_ *harvesterv1.VirtualMachineImage) error {
+func (b *Backend) Delete(vmImg *harvesterv1.VirtualMachineImage) error {
+	targetDVNs := vmImg.ObjectMeta.Namespace
+	targetDVName := vmImg.ObjectMeta.Name
+	_, err := b.dataVolumeClient.Get(targetDVNs, targetDVName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("Target DataVolume (%s/%s) is already gone do no-op found", targetDVNs, targetDVName)
+			return nil
+		}
+		return fmt.Errorf("failed to get DataVolume %s/%s: %v", targetDVNs, targetDVName, err)
+	}
+
+	// delete the DataVolume
+	if err := b.dataVolumeClient.Delete(targetDVNs, targetDVName, &metav1.DeleteOptions{}); err != nil {
+		return fmt.Errorf("failed to delete DataVolume %s/%s: %v", targetDVNs, targetDVName, err)
+	}
 	return nil
 }
 
-func (cb *Backend) AddSidecarHandler() {
+func (b *Backend) AddSidecarHandler() {
+}
+
+func (b *Backend) isDataVolumeCreated(vmImg *harvesterv1.VirtualMachineImage) (bool, error) {
+	targetDVNs := vmImg.ObjectMeta.Namespace
+	targetDVName := vmImg.ObjectMeta.Name
+	_, err := b.dataVolumeClient.Get(targetDVNs, targetDVName, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, fmt.Errorf("failed to get DataVolume %s/%s: %v", targetDVNs, targetDVName, err)
+	}
+	if apierrors.IsNotFound(err) {
+		return false, nil
+	}
+	return true, nil
+}
```

Comment:
Suggest this for readability
```suggestion
	if err != nil {
		if !apierrors.IsNotFound(err) {
			return false, fmt.Errorf("failed to get DataVolume %s/%s: %v", targetDVNs, targetDVName, err)
		}
		return false, nil
	}
	return true, nil
```